File: 0063-unique-paths-ii/0063-unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        
        int[][] dp = new int[m+1][n+1];
        for(int[] arr:dp) Arrays.fill(arr, -1);
        return f(dp, m-1, n-1, obstacleGrid);
    }
    
    int f(int[][]dp, int m, int n, int[][] obstacleGrid){
        if (m<0 || n<0) return 0;
        if (obstacleGrid[m][n] == 1) return 0;
        if (m == 0 && n==0) return 1;
        
        if (dp[m][n] != -1) return dp[m][n];
        
        return dp[m][n] = f(dp, m-1, n, obstacleGrid) + f(dp, m, n-1,obstacleGrid);
    }
}

File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, grid: List[List[int]]) -> int:
        dp = [[-1] * len(grid[0]) for _ in range(len(grid))]
            
        def helper(x, y):
            
            if x == len(grid)-1 and y == len(grid[0])-1 and grid[x][y] == 0: return 1
            if x > len(grid)-1 or y > len(grid[0])-1: return 0
            if grid[x][y] == 1: return 0
            if dp[x][y] != -1: return dp[x][y]

            dp[x][y] = helper(x+1, y) + helper(x, y+1)
            return dp[x][y]
        return helper(0,0)